[WIP] Unit Test for dashboard

proy30 opened the pull request:
This PR introduces the first unit test for the ImpactX dashboard, leveraging the SeleniumBase library to simulate a simulation run by a user with the GUI.

### Key Features
  - The unit test runs a simulation using the parameters from the FODO cell example on the GUI. (takes ~1 minute)
  - The test is considered successful if the simulation is completed without errors.

### Issues

- ~~**Dependency on Running Dashboard:**~~
  - ~~The current implementation requires the ImpactX dashboard to run in a separate terminal before executing the test via `pytest`.~~
  - ~~Can investigate to potentially have dashboard be launched within the test itself.~~
  - ~~Unit test is flaky with passing/failing, unsure if this is due to code error or error in pytest implementation. Need to investigate. Started occurring after the implementation of subprocess.~~
 - **Adding seleniumbase as a dependency**
   - Unsure where to add. (implemented incorrectly)

### Checklist

- [X] Implement unit test for impactx-dashboard (testing with FODO cell simulation parameters.
- [X] Automate the dashboard launch within the test.
- [X] Fix flaky unit test
- [x] Include seleniumbase dependency in CI test files
- [x] Include trame dependencies in CI test files

Resolves #667

proy30 committed:
Added IDs to UI components

proy30 committed:
Added some more IDs to UI components

proy30 committed:
Added 'close' button in lattice settings

Contributes to unit testing by applying a UI component to exit the dialog box and allowing selenium to continue testing the application

proy30 committed:
Removed functionality that converted user input

Initially these were added to convert user input to the correct type in order to be compatible with simulation. However, upon commenting these out for the sake of unit testing compatibility, it was found that the simulation is still functional with user inputs and does not need any conversion of user values. There is potential for underlying problems that have yet to be found.

proy30 committed:
Added unit test for running FODO simulation on dashboard

proy30 commented on tests/python/dashboard/util.py:
```diff
@@ -0,0 +1,22 @@
+        if sb.is_element_present(".trame__loader"):
+            sb.sleep(1)
+        else:
+            print("Ready")
```
This is code directly from https://github.com/Kitware/trame-client/blob/master/trame_client/utils/testing.py. Not sure of the process to credit, if any.

ax3l replied:
Add inline comments with links? :)

proy30 committed:
Added subprocess to start up server with pytest

proy30 committed:
Attempt to fix flaky test

proy30 committed:
Add seleniumbase dependency in CI workflows for testing

proy30 committed:
Added dashboard dependencies in stubs.yml

proy30 committed:
Added missing dashboard dependencies to stubs.yml

proy30 commented on .github/workflows/stubs.yml:
```diff
@@ -41,6 +41,10 @@ jobs:
 
+    - name: Dashboard Dependencies
+      run: |
+        python3 -m pip install -U trame trame-matplotlib trame-plotly trame-router trame-xterm trame-vuetify
```
It seems that adding this caused stub files test to continuously run.

ax3l replied:
fixed

ax3l committed:
Update CI & Requirements Files

ax3l commented on tests/python/dashboard/util.py:
```diff
@@ -0,0 +1,38 @@
+    """
+    Function which starts up impactx-dashboard server.
+    """
+    script_dir = os.path.dirname(os.path.abspath(__file__))
+    working_directory = os.path.join(script_dir, "../../../src/python/impactx")
+    working_directory = os.path.normpath(working_directory)
```
We need to generalize this so it:
- runs in CTests from the build directory
- runs when copied in a package manager with external installs

ax3l commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,74 @@
+    """
+
+    app_process = start_dashboard()
+    time.sleep(10)
```
This is very long.

ax3l commented on tests/python/dashboard/util.py:
```diff
@@ -0,0 +1,38 @@
+    working_directory = os.path.normpath(working_directory)
+
+    return subprocess.Popen(
+        ["python", "-m", "dashboard"],
```
When I run this locally, this literally opens a browser window for me.
Is there a way to run this headless?

ax3l commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,74 @@
+from util import set_input_value, start_dashboard, wait_for_ready
+import time
+
+def test_simulation():
```
```suggestion
def test_dashboard():

```

ax3l committed:
pytest: Graceful Skipping

Mark the test as skipped if `seleniumbase` is not available.

ax3l commented on .github/workflows/stubs.yml:
```diff
@@ -38,8 +38,7 @@ jobs:
         .github/workflows/dependencies/gcc-openmpi.sh
-        python3 -m pip install -U pip setuptools wheel
-        python3 -m pip install -U pip mpi4py pytest pybind11-stubgen pre-commit
+        python3 -m pip install -U pybind11-stubgen
```
```suggestion
        python3 -m pip install -U pybind11-stubgen pre-commit
```

ax3l replied:
done

ax3l committed:
Merge remote-tracking branch 'mainline/development' into selenium

proy30 commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,68 @@
+
+        # Check if "Simulation complete" message is printed
+        xterm_content = sb.get_text("#xterm_component")
+        assert "Simulation complete." in xterm_content
```
This is currently buggy and offering flaky results.

proy30 replied:
Alternative method implemented by having unit test interact with phase space projections, which indirectly mean simulation ran properly.

ax3l commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,82 @@
+    time.sleep(10)
+
+    try:
+        with SB() as sb:
```
Trying if this prevents that the unit test on Linux/macOS opens a browser:
```suggestion
        with SB(headless=True) as sb:

```

ax3l replied:
https://github.com/ECP-WarpX/impactx/pull/672/files#r1717652047

ax3l commented on tests/python/dashboard/test_dashboard.py:
```diff
@@ -0,0 +1,82 @@
+    """
+    from seleniumbase import SB
+
+    app_process = start_dashboard()
```
Should this line go down into the `with SB() as sb:` block?

proy30 committed:
Unit test interacts with phase space projections

This indirectly assures user that sim ran successfully.

proy30 committed:
time adjustments

proy30 committed:
adjust format

proy30 committed:
attempt to run headless

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Fixed UnboundLocalError and PermissionError

changed hardcoded 'python' command to 'sys.executable' and also initialized app_process

proy30 committed:
Replace time.sleep() with alternative

Adjusted function to retry upon failure until timeout has reached max time.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

proy30 committed:
Redid logic to show simulation is complete

ax3l commented:
Continued in #706
